0.2-Elliptic-Curve-Math.ipynb

```diff
@@ -0,0 +1,247 @@
+    "\n",
+    "In the example below, addition is performed with both integers and the `ECKey` class, and evaluated for equality."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# int() operations\n",
+    "a = random.randrange(1, SECP256K1_ORDER)\n",
+    "b = random.randrange(1, SECP256K1_ORDER)\n",
+    "ab = (a + b) % SECP256K1_ORDER\n",
+    "\n",
+    "# ECkey(operation)\n",
+    "a_key = ECKey()\n",
+    "a_key.set(a.to_bytes(32,'big'), True) \n",
+    "ab_key = a_key.add(b.to_bytes(32,'big'))\n",
+    "\n",
+    "# Ensure equivalency.\n",
+    "print(int().from_bytes(ab_key.get_bytes(),'big') == ab)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### 1.2 Example: Scalar Multiplication over SECP256K1 order\n",
+    "\n",
+    "In the example below, multiplication is performed with both integers and the `ECKey` class, and evaluated for equality."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# int() operations\n",
+    "a = random.randrange(1, SECP256K1_ORDER)\n",
+    "b = random.randrange(1, SECP256K1_ORDER)\n",
+    "ab = (a * b) % SECP256K1_ORDER\n",
+    "\n",
+    "# ECkey(operation)\n",
+    "a_key = ECKey()\n",
+    "a_key.set(a.to_bytes(32,'big'), True) \n",
+    "ab_key = a_key.mul(b.to_bytes(32,'big'))\n",
+    "\n",
+    "# Ensure operations are equivalent.\n",
+    "print(int().from_bytes(ab_key.get_bytes(),'big') == ab)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### 1.3 _Programming Exercise:_ Distributivity of scalar operations\n",
+    "\n",
+    "In this exercise we wish to demonstrate the distributivity property of scalar addition and multiplication, whilst getting familiarized with both integer modulo operations and the private key `ECKey` methods.\n",
+    "\n",
+    "Consider: `(a - b) * c = a * c - b * c` over SECP256k1\n",
+    "* Demonstrate that the equation holds\n",
+    "* Compute the left side with the Python `int` class.\n",
+    "* Compute the right side with the Bitcoin Core `ECKey` class.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "a = random.randrange(1, SECP256K1_ORDER)\n",
+    "b = random.randrange(1, SECP256K1_ORDER)\n",
+    "c = random.randrange(1, SECP256K1_ORDER)\n",
+    "\n",
+    "# Left: Compute (a - b) * c\n",
+    "\n",
+    "\n",
+    "# Right: Compute -b * c\n",
+    "\n",
+    "\n",
+    "# Right: Compute a * c\n",
+    "\n",
+    "\n",
+    "# Right: Compute a * c - b * c\n",
+    "\n",
+    "\n",
+    "# Left/Right: Assert equality\n",
```

Comment:
nit: fill in this line already, so the students just need to make it pass:

`print(int().from_bytes(right.get_bytes(),'big') == left)` or whatever

Reply:
fixed